### DISCUSSION TITLE
[Feature Request] [form] indication whether form changes after prefill

### DISCUSSION THREAD

**Discussion by LarsDenBakker** [ing-bank/lion#552] [General]
When submitting a form, we sometimes need to take a different action based on whether the user actually made any changes to the form. When we find a solution for https://github.com/ing-bank/lion/issues/547 we can use the dirty/prefilled state for this.

A more advanced scenario is when a user made some changes, but then undos their changes and the form is back to the initial state it started in. It would be useful to be able to detect this state.

Would this be possible to implement?

**Comment by tlouisse** [comment]
> It would be useful to be able to detect this state.

I assume that after the user manually undos changes, the form will still be in **dirty** state?
Is this what you are looking for? Or you want it a check to see whether the end state is exactly the begin state? 

We do store a protected `_initalModelValue`.  You could do a deep comparison on the current values and those. Or maybe we should come up with a new interaction state?

**Comment by LarsDenBakker** [comment]
Yes the form is still in a dirty state, I think that's correct. But it would be useful to have a way to detect whether there was any change in the input values from the prefilled state. This comparison can get pretty tricky, but each form element could have a different comparison method.

**Comment by tlouisse** [comment]
Alright. Pseudo code would probable be:

```js
// InteractionStateMixin + FormGroup 
get hasValueChangeSinceInit() {
  return deepComparison(this._initialModelValue, this.modelValue);
}
```

**Comment by LarsDenBakker** [comment]
do you think this will be possible from a form level? maybe it should delegate comparison to the fields like the other properties? then each field can decide how the comparison should happen

**Comment by tlouisse** [comment]
It works on every FormControl (== fieldset/field). _initalModelValue is already delegated:
```js
  get _initialModelValue() {
    return this._getFromAllFormElements('_initialModelValue');
  }
```
https://github.com/ing-bank/lion/blob/master/packages/fieldset/src/LionFieldset.js#L439

But of course you can delegate on the Fieldset (from which the LionForm extends) like:

```js
 return this._getFromAllFormElements('hasValueChangeSinceInit');
```

**Comment by jorenbroekema** [comment]
I like the idea, I would go with `hasValueChanged` as a name for the getter. The init part is I think redundant since if it "has changed", it will be different from "init", no way around that I think?

Closing as enhancement with votes needed. Our workflow with enhancements is that we close them and have people vote on them, then we sort our enhancements by most votes to see which ones should be prioritized: https://github.com/ing-bank/lion/issues?q=is%3Aissue+is%3Aclosed+label%3Aenhancement+sort%3Areactions-%2B1-desc
